test/Gen.hs

```diff
@@ -125,52 +143,72 @@ class Forall g f | g -> f, f -> g where
 
-instance Eq a => Eq (With a) where
-  (==) = (==) `on` getWith
+atom :: String -> a -> Gen a
+atom s = Gen . pure . Pure (const (showString s))
 
-instance Applicative With where
-  pure = atom "_"
-  With' sf f <*> With' sa a = With' (\ d -> showParen (d > 10) (sf 10 . showString " " . sa 11)) (f a)
+label :: String -> a -> Gen a
+label s = addLabel s . atom s
 
-instance Show (With a) where
-  showsPrec d (With' s _) = s d
+infixL :: Int -> String -> (a -> b -> c) -> Gen (a -> b -> c)
+infixL p s f = Gen (pure (InfixL p s f))
 
-showing :: Show a => a -> With a
-showing = With' . flip showsPrec <*> id
+infixR :: Int -> String -> (a -> b -> c) -> Gen (a -> b -> c)
+infixR p s f = Gen (pure (InfixR p s f))
 
-showingFn :: (Show a, Show b) => Fn a b -> With (a -> b)
-showingFn = With' . flip showsPrec <*> apply
+addLabel :: String -> Gen a -> Gen a
+addLabel s = Gen . (>>= \ a -> a <$ tell (Set.singleton (fromString s))) . runGen
 
-atom :: String -> a -> With a
-atom s = With' (\ _ -> showString s)
 
-liftWith :: String -> (a -> b) -> With a -> With b
-liftWith s w a = atom s w <*> a
+data Term a where
+  Pure :: (Int -> ShowS) -> a -> Term a
+  InfixL :: Int -> String -> (a -> b -> c) -> Term (a -> b -> c)
+  InfixR :: Int -> String -> (a -> b -> c) -> Term (a -> b -> c)
+  (:<*>) :: Term (a -> b) -> Term a -> Term b
 
-liftWith2 :: String -> (a -> b -> c) -> With a -> With b -> With c
-liftWith2 s w a b = atom s w <*> a <*> b
+infixl 4 :<*>
 
+runTerm :: Term a -> a
+runTerm = \case
+  Pure _ a -> a
+  InfixL _ _ f -> f
+  InfixR _ _ f -> f
+  f :<*> a -> runTerm f $ runTerm a
 
-pattern With :: a -> With a
-pattern With a <- (getWith -> a)
+instance Functor Term where
+  fmap = liftA
 
-{-# COMPLETE With #-}
+instance Applicative Term where
+  pure = Pure (const (showString "_"))
+  (<*>) = (:<*>)
 
-pattern Fn :: (a -> b) -> Fn a b
-pattern Fn f <- (apply -> f)
+instance Show (Term a) where
+  showsPrec d = \case
+    Pure s _ -> s d
+    InfixL _ s _ -> showParen True (showString s)
+    InfixR _ s _ -> showParen True (showString s)
+    InfixL p s _ :<*> a :<*> b -> showParen (d > p) (showsPrec p a . showString " " . showString s . showString " " . showsPrec (succ p) b)
+    InfixR p s _ :<*> a :<*> b -> showParen (d > p) (showsPrec (succ p) a . showString " " . showString s . showString " " . showsPrec p b)
+    InfixL p s _ :<*> a -> showParen True (showsPrec p a . showString " " . showString s)
+    InfixR p s _ :<*> a -> showParen True (showsPrec (succ p) a . showString " " . showString s)
+    f :<*> a -> showParen (d > 10) (showsPrec 10 f . showString " " . showsPrec 11 a)
 
-{-# COMPLETE Fn #-}
 
-pattern FnWith :: (a -> b) -> Fn a (With b)
-pattern FnWith f <- (fmap getWith . apply -> f)
+newtype Gen a = Gen { runGen :: WriterT (Set.Set LabelName) Hedgehog.Gen (Term a) }
+  deriving (Functor)
 
-{-# COMPLETE FnWith #-}
+instance Applicative Gen where
+  pure = Gen . pure . pure
+  Gen m1 <*> Gen m2 = Gen ((<*>) <$> m1 <*> m2)
```

Comment:
Note that we don’t offer a `Monad` instance for `Gen`—I haven’t figured out whether we can have good support for `Term`, and if so, how.